### PR TITLE
Extra loggers

### PR DESCRIPTION
**Pull request by petebacondarwin**
These two new packages, which should end up in dgeni-packages, add the ability to do more detailed logging of what is going on inside Dgeni during document processing.

They rely upon the event hooks that were added in #133.

There is no documentation of the loggers yet - just wanted to get some feedback on the idea.

### CODE REVIEW

File: lib/packages/docDiffLogger.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+var _ = require('lodash');
+var Package = require('../Package');
+var util = require('util');
```
Comment (forresst) [inline lib/packages/docDiffLogger.js]: util : unused

File: lib/packages/docDiffLogger.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+var diff = require('objectdiff').diff;
+var firstDocs, startDocs, lastDocs;
+var firstProcessor, startProcessor, lastProcessor;
```
Comment (forresst) [inline lib/packages/docDiffLogger.js]: irstProcessor, startProcessor, lastProcessor : What are these three variables? They are assigned but not used, I do not see their utilities

File: lib/packages/docDiffLogger.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+.factory('docDiffLoggerOptions', function() {
+  return options;
+})
+
+.eventHandler('processorStart', function() {
+  return function capturePreviousDocs(event, processor, docs) {
+    firstProcessor = firstProcessor || processor;
+    firstDocs = firstDocs || _.cloneDeep(docs);
+
+    if ( options.start == processor.name ) {
```
Comment (forresst) [inline lib/packages/docDiffLogger.js]: === ?

File: lib/packages/docDiffLogger.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    if ( options.start && !startDocs ) {
+      throw new Error('docDiffLogger: missing start processor');
+    }
+    if ( options.end && !endDocs ) {
+      throw new Error('docDiffLogger: missing end processor');
+    }
+    if ( !options.end ) {
+      logDiff(log);
+    }
+  }
```
Comment (forresst) [inline lib/packages/docDiffLogger.js]: miss `;`

File: lib/packages/docDiffLogger.spec.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        };
+      })
+      .processor('first', function() {
+        return {
+          $runAfter: ['initial'],
+          $process: function(docs) {
+            docs.push({ id: 'doc-1' });
+            docs.push({ id: 'doc-2' });
+          }
+        }
```
Comment (forresst) [inline lib/packages/docDiffLogger.spec.js]: miss `;`

File: lib/packages/docDiffLogger.spec.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        };
+      });
+  });
+
+  it("should log the difference between the first and last processor", function(done) {
+    dgeni.generate()
+      .then(function() {
+        expect(mockLogger.info).toHaveBeenCalledWith(jasmine.objectContaining({ changed: 'object change' }));
+        done();
+      })
```
Comment (forresst) [inline lib/packages/docDiffLogger.spec.js]: miss ;

File: lib/packages/docDiffLogger.spec.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        console.log(error.stack);
+        done(error);
+      });
+  });
+
+  it("should log the difference between the start and last processor", function(done) {
+    dgeni.package('testConfig')
+      .config(function(docDiffLoggerOptions) {
+        docDiffLoggerOptions.start = 'first';
+      })
```
Comment (forresst) [inline lib/packages/docDiffLogger.spec.js]: miss ;

File: lib/packages/docDiffLogger.spec.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        console.log(error.stack);
+        done(error);
+      });
+  });
+
+  it("should log the difference between the first and end processor", function(done) {
+    dgeni.package('testConfig')
+      .config(function(docDiffLoggerOptions) {
+        docDiffLoggerOptions.end = 'first';
+      })
```
Comment (forresst) [inline lib/packages/docDiffLogger.spec.js]: miss ;

File: lib/packages/docDiffLogger.spec.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        done(error);
+      });
+  });
+
+  it("should log the difference between the start and end processor", function(done) {
+    dgeni.package('testConfig')
+      .config(function(docDiffLoggerOptions) {
+        docDiffLoggerOptions.start = 'second';
+        docDiffLoggerOptions.end = 'second';
+      })
```
Comment (forresst) [inline lib/packages/docDiffLogger.spec.js]: miss ;

File: lib/packages/trackDocLogger.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
+var _ = require('lodash');
+var Package = require('../Package');
+var diff = require('objectdiff').diff;
```
Comment (forresst) [inline lib/packages/trackDocLogger.js]: diff : unused

File: lib/packages/trackDocLogger.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+     }
+    }
+  };
+})
+
+.eventHandler('generationEnd', function(log) {
+  return function() {
+    log.info('trackDocLogger settings:', options);
+    log.info('trackDocLogger tracked changes:', generations);
+  }
```
Comment (forresst) [inline lib/packages/trackDocLogger.js]: miss `;`

File: lib/packages/trackDocLogger.spec.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            ];
+          }
+        };
+      })
+      .processor('first', function() {
+        return {
+          $runAfter: ['initial'],
+          $process: function(docs) {
+          }
+        }
```
Comment (forresst) [inline lib/packages/trackDocLogger.spec.js]: miss `;`

**Comment by petebacondarwin** [comment]
Merged

**Pull request closed without merging**